source/developer/interactive-message-buttons.rst

```diff
@@ -0,0 +1,116 @@
+
+Interactive Message Buttons (Beta)
+===================================
+
+Mattermost supports interactive message buttons for `incoming <https://docs.mattermost.com/developer/webhooks-incoming.html>`_ and `outgoing webhooks <https://docs.mattermost.com/developer/webhooks-outgoing.html>`_, and for `custom slash commands <https://docs.mattermost.com/developer/slash-commands.html>`_ via actions. They help make your integrations richer by completing common tasks inside Mattermost covnersations, increasing user engagement and productivity.
+
+Use these buttons to simplify complex workflows by allowing users to take quick actions directly through your integration post. For example, the buttons enable your integration to:
+
+- update sales opportunities in your CRM system from Mattermost
+- conduct a customer survey or a poll
+- file a report on market trends
+
+Interactive message buttons are not yet supported on Mattermost mobile apps.
+
+// XXX Add GIF image (AB preparing one for marketing)
+
+.. toctree::
+   :maxdepth: 2
+
+Button Options
+---------------
+
+Add message buttons as ``actions`` in your integration `message attachments <https://docs.mattermost.com/developer/message-attachments.html>`_. Each attachment can contain up to XXX actions. // XXX @ccbrown how many?
+
+The following payload gives an example webhook that uses message buttons.
+
+.. code-block:: text
+
+  {
+    "attachments": [
+      {
+        "pretext": "This is the attachment pretext.",
+        "text": "This is the attachment text.",
+        "actions": [
+          {
+            "name": "Ephemeral Message",
+            "integration": {
+              "url": "http://127.0.0.1:7357",
+              "context": {
+                "action": "ephemeral"
+              }
+            }
+          }, {
+            "name": "Update",
+            "integration": {
+              "url": "http://127.0.0.1:7357",
+              "context": {
+                "action": "update"
+              }
+            }
+          }
+        ]
+      }
+    ]
+  }
+
+The integration can respond with an update to the original post, or with an ephemeral message:
+
+.. code-block:: text
+
+  {
+    "update": {
+      "message": "Updated!"
+    },
+    "ephemeral_text": "You updated the post!"
+  }
+
+// XXX Insert an image here
+
+Below we give a brief description of each parameter to help you customize the webhook post in Mattermost. For more information on message attachments, `see our documentation <https://docs.mattermost.com/developer/message-attachments.html>`_.
+
+Name
+  Give your action a descriptive name.
+
+URL
+  The actions are backed by an integration that handles HTTP POST requests when users clicks the message button. The URL parameter determines where this action is sent to. The request contains an ``application/x-www-form-urlencoded`` JSON string. // XXX @ccbrown is this correct?
+
+Context // XXX @ccbrown Need a bit of help to more specifically explain this parameter.
```

Comment:
There are two things that virtually every integration will need to use this for:

1. Identifying which action was triggered: For example, a GitHub integration might store something like this in the context:

    ```json
    {
        "repo": "mattermost/mattermost-server",
        "pr": 1234,
        "action": "merge"
    }
    ```

    Then when the button is clicked and it receives that context back, the integration knows that the intention was to merge the pull request identified by the context.

2. Authenticating the server: If your integration isn't behind a firewall, anyone can forge requests that look like they came from the Mattermost server. One very important property of the context is that the entire context is kept confidential, so it's safe to put secrets in it if needed. Users won't be able to see them. So if your integration isn't behind a firewall, you might add a token to your context:

    ```json
    {
        "repo": "mattermost/mattermost-server",
        "pr": 1234,
        "action": "merge",
        "token": "somerandomlygeneratedsecret"
    }
    ```

    Then, when you receive the request, you can verify that the token matches one that you previously generated and know that the request is legitimately coming from the Mattermost server.

    Depending on the application, integrations may be able to perform authentication statelessly with cryptographic signatures:

    ```json
    {
        "repo": "mattermost/mattermost-server",
        "pr": 1234,
        "action": "merge",
        "signature": "mycryptographicsignature"
    }
    ```

It's also possible for integrations to do both of these things with a single token and simple store something like this as context:

```json
{
    "action_id": "someunguessableactionid"
}
```

Then when the integration gets the request, it can look up what to do based on the action id. But this requires integrations to maintain a bit more state on their end.